lib/actions-oclif/preload.ts

```diff
@@ -0,0 +1,552 @@
+			options.commit,
+			options.image,
+			options.splashImage,
+			options.proxy, // TODO: Currently always undefined, investigate approach in ssh command.
+			options.dontCheckArch,
+			options.pinDevice,
+			certificates,
+		);
+
+		let gotSignal = false;
+
+		nodeCleanup(function (_exitCode, signal) {
+			if (signal) {
+				gotSignal = true;
+				nodeCleanup.uninstall(); // don't call cleanup handler again
+				preloader.cleanup().then(() => {
+					// calling process.exit() won't inform parent process of signal
+					process.kill(process.pid, signal);
+				});
+				return false;
+			}
+		});
+
+		if (process.env.DEBUG) {
+			preloader.stderr.pipe(process.stderr);
+		}
+
+		preloader.on('progress', progressHandler);
+		preloader.on('spinner', spinnerHandler);
+
+		try {
+			await new Promise((resolve, reject) => {
+				preloader.on('error', reject);
+				resolve(this.prepareAndPreload(preloader, balena, options));
+			});
+		} catch (err) {
+			if (instanceOf(err, balena.errors.BalenaError)) {
+				const code = err.code ? `(${err.code})` : '';
+				throw new ExpectedError(`${err.message} ${code}`);
+			} else {
+				throw err;
+			}
+		} finally {
+			if (!gotSignal) {
+				await preloader.cleanup();
+			}
+		}
+	}
+
+	readonly applicationExpandOptions: ResourceExpand = {
+		owns__release: {
+			$select: ['id', 'commit', 'end_timestamp', 'composition'],
+			$orderby: [{ end_timestamp: 'desc' }, { id: 'desc' }],
+			$expand: {
+				contains__image: {
+					$select: ['image'],
+					$expand: {
+						image: {
+							$select: ['image_size', 'is_stored_at__image_location'],
+						},
+					},
+				},
+			},
+			$filter: {
+				status: 'success',
+			},
+		},
+		should_be_running__release: {
+			$select: 'commit',
+		},
+	};
+
+	allDeviceTypes: DeviceTypeJson.DeviceType[];
+	async getDeviceTypes() {
+		if (this.allDeviceTypes !== undefined) {
+			return this.allDeviceTypes;
+		}
+		const balena = getBalenaSdk();
+		return balena.models.config
+			.getDeviceTypes()
+			.then((deviceTypes) => _.sortBy(deviceTypes, 'name'))
+			.then((deviceTypes) => {
+				this.allDeviceTypes = deviceTypes;
+				return deviceTypes;
+			});
+	}
+
+	isCurrentCommit(commit: string) {
+		return commit === 'latest' || commit === 'current';
+	}
+
+	getDeviceTypesWithSameArch(deviceTypeSlug: string) {
+		return this.getDeviceTypes().then((deviceTypes) => {
+			const deviceType = _.find(deviceTypes, { slug: deviceTypeSlug });
+			if (!deviceType) {
+				throw new Error(
+					`Device type "${deviceTypeSlug}" not found in API query`,
+				);
+			}
+			return _(deviceTypes)
+				.filter({ arch: deviceType.arch })
+				.map('slug')
+				.value();
+		});
+	}
+
+	getApplicationsWithSuccessfulBuilds(deviceTypeSlug: string) {
+		const balena = getBalenaSdk();
+
+		return this.getDeviceTypesWithSameArch(deviceTypeSlug).then(
+			(deviceTypes) => {
+				const options: PineOptions<
+					ApplicationWithDeviceType & { should_be_running__release: [Release?] }
+				> = {
```

Comment:
I think that this should have been:
```ts
const options: PineOptions<Application> = {
```
for which to work you should also type `applicationExpandOptions` as `BalenaSdk.PineExpand<Application>`

Reply:
Ok thanks, will add that in later PR.

Reply:
Ah, I can't use `BalenaSdk.PineExpand<Application>` at the moment because yesterday page added a fix to the typings for the `$orderby` format used here.

Reply:
I'm already on it @srlowe

Reply:
@srlowe 
See: https://github.com/balena-io/balena-cli/pull/2011

This is my current proposal on how to type everything with the current SDK typings.